core/plugins/org.polarsys.capella.core.data.business.queries/src/org/polarsys/capella/core/business/queries/queries/cs/GetAvailable_PhysicalPath_AllocatedComponentExchanges.java

```diff
@@ -14,98 +14,80 @@
-		List<Component> terminalPCs = new ArrayList<Component>();
-		for (PhysicalPathInvolvement first : PhysicalPathExt.getFlatFirstPhysicalPathInvolvments(element)) {
-			AbstractPathInvolvedElement involvedElt = first.getInvolvedElement();
-			if (involvedElt instanceof Part) {
-				AbstractType pc = ((Part) involvedElt).getAbstractType();
-				if ((pc instanceof Component) && !terminalPCs.contains(pc)) {
-					terminalPCs.add((Component) pc);
-				}
+	private Set<ComponentExchange> getPhysicalPathAllocatedComponentExchanges(PhysicalPath element) {
+		Set<ComponentExchange> availableElements = new HashSet<>();
+		List<Component> terminalPCs = new ArrayList<>();
+		
+		retreiveTerminalPCs(PhysicalPathExt.getFlatFirstPhysicalPathInvolvments(element), terminalPCs);
+		retreiveTerminalPCs(PhysicalPathExt.getFlatLastPhysicalPathInvolvments(element), terminalPCs);
+
+		for(int i = 0; i<terminalPCs.size()-1; i++) {
+			Component sourceComponent = terminalPCs.get(i);
+			for(int j = i+1; j<terminalPCs.size(); j++) {
+				Component targetComponent = terminalPCs.get(j);
+				availableElements.addAll(getAvailableExchanges(sourceComponent, targetComponent));
 			}
 		}
-		for (PhysicalPathInvolvement last : PhysicalPathExt.getFlatLastPhysicalPathInvolvments(element)) {
-			AbstractPathInvolvedElement involvedElt = last.getInvolvedElement();
+		return availableElements;
+	}
+	
+  private void retreiveTerminalPCs(Collection<PhysicalPathInvolvement> physicalPathInvolvments, List<Component> terminalPCs) {
+		for (PhysicalPathInvolvement first : physicalPathInvolvments) {
+			AbstractPathInvolvedElement involvedElt = first.getInvolvedElement();
 			if (involvedElt instanceof Part) {
 				AbstractType pc = ((Part) involvedElt).getAbstractType();
 				if ((pc instanceof Component) && !terminalPCs.contains(pc)) {
 					terminalPCs.add((Component) pc);
 				}
 			}
 		}
-		for (Component firstPC : terminalPCs) {
-			for (Component lastPC : terminalPCs) {
-				if (!firstPC.equals(lastPC)) {
-					for (CapellaElement exchange : getAvailableExchanges(firstPC, lastPC)) {
-						if (!availableElements.contains(exchange)) {
-							availableElements.add(exchange);
-						}
-					}
-				}
-			}
-		}
-		return availableElements;
 	}
-	
-  protected List<CapellaElement> getAvailableExchanges(Component sourceComponent, Component targetComponent) {
-    List<CapellaElement> availableElements = new ArrayList<CapellaElement>();
 
-    // collect deployed component in any of the source part
-    List<Component> sourceDeployedElements =  new ArrayList<Component>(1);
-    // collect deployed component in any of the target part
-    List<Component> targetDeployedElements =  new ArrayList<Component>(1);
+private Set<ComponentExchange> getAvailableExchanges(Component sourceComponent, Component targetComponent) {
 
     // all the descendants of the @sourceComponent
     Collection<Component> sourceComponents = ComponentExt.getAllSubUsedComponents(sourceComponent);

```

Comment:
I think PartExt.getComponentsOfParts(ComponentExt.getAllSubUsedParts(component, true)) that will do like below.